Add support for bound function arguments (#291)
dfinity/candid: spec/Candid.md

<reftype> ::= func <functype> | ...
```

A function reference may be represented by a simple closure, encapsulating a prefix of arguments that have previously been bound. These hidden arguments do not appear in the function type, but will be forwarded implicitly when the function itself is invoked.

##### Examples

```
type engine = service {
  search : (query : text, callback : func (vec result) -> ());
}
```

```
type wallet = service {
  topup : (amount : nat) -> ();

>   forward : (call : () -> ()) -> ();

Reviewer: ```suggestion
  forward : (call : func () -> ()) -> ();
```
syntactically, how do we know if the function has bound arguments? Does the closure supports return type?

Author: You mean, know it by the type? We don't, it's always allowed. Do you think that would cause issues for bindings?

Reviewer: I see. The existing example has a `callback` function, which can have bound arguments as well. You are adding another example to illustrate the use of bound arguments, which makes me think there may be syntactic difference for the bound args.

So any function reference can have bound arguments. The return value will be decoded according to the function signature, right? That means we need an abstraction for argument tuples. And if we use the `dynamic` type as return type, `dynamic` represents argument tuples instead of a single argument.

Reviewer 2: > That means we need an abstraction for argument tuples.

Yep.  That's my thinking too, and more agreement in the comments for #292

<fieldval> ::= <nat> = <annval>

<refval> ::=

>   | principal <text>                      (principal URI)

Reviewer: `principal` is now a primitive type, so we can move this to `<primval>`

Author: Done.

C[principal <: principal](principal <text>) = principal <text>
```

However, functions can be converted into closures with an empty list of bound arguments:
```

> C[func <functype> <: closure <functype'>](func <text>.<id>) = clos(func <text>.<id>, .)

Author: Ah, yes.

Likewise, there are two forms of Candid values for function references:

* `ref(r)` indicates an opaque reference, understood only by the underlying system.

> * `pub(s,n,v*:t*)`, indicates the public method name `n` of the service referenced by `s`, possibly followed by a list of type-annotated bound argument values.

Reviewer 3: You don’t want to support binding arguments to references?

Why not make it a recursive definition that allows you to bind arguments to any function reference (whether it’s opaque, public, or itself a closure – after all, these are all values of the same type, so binding to them should be allowed).
```suggestion
* `closure(f,v*:t*)`, indicates the function reference `f`, followed by a list of type-annotated bound argument values.
```
```

T(func (<datatype1>*) -> (<datatype2>*) <funcann>*) =
  sleb128(-22) T*(<datatype1>*) T*(<datatype2>*) T*(<funcann>*) // 0x6a
T(service {<methtype>*}) =
  sleb128(-23) T*(<methtype>*)                                  // 0x69

> T(closure (<datatype1>*) -> (<datatype2>*) <funcann>*) =

Author: Done.

M(ref(r) : service <actortype>) = i8(0)
M(id(v*) : service <actortype>) = i8(1) M(v* : vec nat8)

M(ref(r)         : func <functype>) = i8(0)
M(pub(s,n)       : func <functype>) = i8(1) M(s : service {}) M(n : text)
M(pub(s,n,v*:t*) : func <functype>) = i8(2) M(s : service {}) M(n : text) M*(v* : t*)

M(ref(r) : principal) = i8(0)
M(id(v*) : principal) = i8(1) M(v* : vec nat8)

M* : <val>* -> <datatype>* -> i8*

> M*(v^N : <datatype>^N) = leb128(N) M(v : <datatype>)^N

Reviewer: this will be a breaking change? or the bound argument is stored in `R` which was not used before?

Author: No, R is only used if there are reference types that need it, as before. The type is determined from M.

But I think this is a breaking change in so far that a receiver not understanding the new encoding of closures would choke on it. The only way to avoid this is by indeed introducing a separate closure type as a future type, it seems. Hm, that would be ugly...

Reviewer 3: A new type seems to be appropriate. Not breaking existing stuff is part of our value proposition, and given the complexity of this maybe it’s good if a service can say that they really only support plain function references, but not closures?

> M*(v^N : <datatype>^N) = leb128(N) M(v : <datatype>)^N

Reviewer 3: Where is the type of `v` encoded? I’d expect something like
```
 M*(v^N : <datatype>^N) = leb128(N) I(<datatype>^N)^N  M(v : <datatype>)^N
```
to match what we do in `B`

M(ref(r) : service <actortype>) = i8(0)
M(id(v*) : service <actortype>) = i8(1) M(v* : vec nat8)

M(ref(r)        : func <functype>) = i8(0)
M(pub(s,n)      : func <functype>) = i8(1) M(s : service {}) M(n : text)
M(clos(f,v*:t*) : closure <functype>) =
  FM(
    i8(2) M(f : func <functype>) TM*(v* : t*),
    R(f : func <functype>) R*(v* : t*)
  )

M(ref(r) : principal) = i8(0)
M(id(v*) : principal) = i8(1) M(v* : vec nat8)

TM : <val> -> <datatype> -> i8*
TM(v : <datatype>) = I(<datatype>) M(v : <datatype>)

TM* : <val>* -> <datatype>* -> i8*
TM*(v^N : <datatype>^N) = leb128(N) TM(v : <datatype>)^N
```

> The meta-function `FM` constructs a backwards-compatible encoding for values of for [future types](#deserialisation-of-future-types).

Reviewer 4: ```suggestion
The meta-function `FM` constructs a backwards-compatible encoding for values [future types](#deserialisation-of-future-types).
```

Author: ```suggestion
The meta-function `FM` constructs a backwards-compatible encoding for values of [future types](#deserialisation-of-future-types).
```

M(ref(r) : service <actortype>) = i8(0)
M(id(v*) : service <actortype>) = i8(1) M(v* : vec nat8)

M(ref(r)        : func <functype>) = i8(0)
M(pub(s,n)      : func <functype>) = i8(1) M(s : service {}) M(n : text)
M(clos(f,v*:t*) : closure <functype>) =
  leb128(|i8(2) M(f : func <functype>) TM*(v* : t*)|)
  leb128(|R(f : func <functype>) R*(v* : t*)|)

>   i8(2) M(f : func <functype>) TM*(v* : t*)

Author: Done'

M(ref(r) : principal) = i8(0)
M(id(v*) : principal) = i8(1) M(v* : vec nat8)

TM : <val> -> <datatype> -> i8*

> TM(v : <datatype>) = T(<datatype>) M(v : <datatype>)

Author: Good point, changed.

Deserialisation uses the following mechanism for robustness towards future extensions:

* A serialised type may be headed by an other than -1 to -24 . Any such opcode is followed by an LEB128-encoded count, and then a number of bytes corresponding to this count. A type represented that way is called a *future type*.
  ```
  FT : i32 -> i8* -> i8*
  FT(n, b*) = sleb128(n) leb128(|b*|) b*
  ```

* A value corresponding to a future type is called a *future value*. It is represented by two LEB128-encoded counts, *m* and *n*, followed by a *m* bytes in the memory representation M and accompanied by *n* corresponding references in R.
  ```
  FM : i8* -> ref* -> i8*

>   FM(b*, r*) = leb128(|b*|) leb128(r*) b*

Reviewer 3: ```suggestion
  FM(b*, r*) = leb128(|b*|) leb128(|r*|) b*
```
